src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,494 @@
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressedField);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testFieldsInStatelessChecksShouldBeImmutable() {
+        final DescribedPredicate<JavaField> moduleProperties = new ModulePropertyPredicate();
+
+        final ArchCondition<JavaField> beSuppressedField = new SuppressionArchCondition<>(
+            SUPPRESSED_FIELDS_IN_MODULES, "be suppressed");
+
+        final ArchCondition<JavaField> bePrivateExceptMessageKeys =
+            new ArchCondition<>("be private") {
+
+                @Override
+                public void check(JavaField item, ConditionEvents events) {
+                    if (!item.getName().startsWith("MSG_")
+                        && !item.getModifiers().contains(JavaModifier.PRIVATE)) {
+                        final String message = String
+                            .format(Locale.ROOT, "Field <%s> should %s in %s",
+                                    item.getFullName(), getDescription(),
+                                    item.getSourceCodeLocation().toString());
+                        events.add(SimpleConditionEvent.violated(item, message));
+                    }
+                }
+            };
+
+        final ArchRule fieldsInStatelessChecksShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .areAnnotatedWith(StatelessCheck.class)
+            .and(are(not(moduleProperties)))
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould(bePrivateExceptMessageKeys)
+            .orShould(beSuppressedField);
+
+        fieldsInStatelessChecksShouldBeImmutable.check(CHECKSTYLE_CHECKS);
+    }
+
+    /**
+     * Test to ensure classes with immutable fields are annotated with {@link StatelessCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testClassesWithImmutableFieldsShouldBeStateless() {
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition<>(
+            SUPPRESSED_CLASSES_FOR_STATELESS_CHECK_RULE, "be suppressed");
+
+        final ArchRule classesWithImmutableFieldsShouldBeStateless = classes()
+            .that(have(IMMUTABLE_FIELDS))
+            .and()
+            .haveSimpleNameEndingWith("Check")
```

Comment:
Similar to a previous item, we can't use `ModuleReflectionUtil#isCheckstyleModule`?

Reply:
Removed.